src/interface/variable_pack.hpp

```diff
@@ -158,203 +255,181 @@ struct PackAndIndexMap {
-          if (ndim >= 2) host_f2(vindex) = v->flux[X2DIR].Get(k, j, i);
-          if (ndim >= 3) host_f3(vindex) = v->flux[X3DIR].Get(k, j, i);
+          if (v->IsAllocated()) {
+            host_f1(vindex) = v->flux[X1DIR].Get(k, j, i);
+            if (ndim >= 2) host_f2(vindex) = v->flux[X2DIR].Get(k, j, i);
+            if (ndim >= 3) host_f3(vindex) = v->flux[X3DIR].Get(k, j, i);
+          } else {
+            host_f1(vindex) = ParArray3D<T>("unallocated_f1_" + v->label(), 0, 0, 0);
+            if (ndim >= 2) {
+              host_f2(vindex) = ParArray3D<T>("unallocated_f2_" + v->label(), 0, 0, 0);
+            }
+            if (ndim >= 3) {
+              host_f3(vindex) = ParArray3D<T>("unallocated_f3_" + v->label(), 0, 0, 0);
+            }
+          }
+
           vindex++;
         }
       }
     }
-    if (vmap != nullptr) {
-      vmap->insert(
+
+    if (vmap_out != nullptr) {
+      vmap_out->insert(
           std::pair<std::string, IndexPair>(v->label(), IndexPair(vstart, vindex - 1)));
     }
   }
-  if (vmap != nullptr && sparse_name != "") {
-    vmap->insert(std::pair<std::string, IndexPair>(sparse_name,
-                                                   IndexPair(sparse_start, vindex - 1)));
-  }
 
-  Kokkos::deep_copy(f1, host_f1);
-  Kokkos::deep_copy(f2, host_f2);
-  Kokkos::deep_copy(f3, host_f3);
+  Kokkos::deep_copy(f1_out, host_f1);
+  Kokkos::deep_copy(f2_out, host_f2);
+  Kokkos::deep_copy(f3_out, host_f3);
 }
 
 template <typename T>
-VariableFluxPack<T> MakeFluxPack(const vpack_types::VarList<T> &vars,
-                                 const vpack_types::VarList<T> &flux_vars,
-                                 PackIndexMap *vmap = nullptr) {
+VariableFluxPack<T> MakeFluxPack(const CellVariableVector<T> &vars,
+                                 const CellVariableVector<T> &flux_vars,
+                                 PackIndexMap *vmap_out = nullptr) {
   // count up the size
   int vsize = 0;
   for (const auto &v : vars) {
-    vsize += v->GetDim(6) * v->GetDim(5) * v->GetDim(4);
+    // we also count unallocated vars because the total size needs to be uniform across
+    // meshblocks that meshblock packs will work
+    vsize += v->NumComponents();
   }
   int fsize = 0;
   for (const auto &v : flux_vars) {
-    fsize += v->GetDim(6) * v->GetDim(5) * v->GetDim(4);
+    // we also count unallocated vars because the total size needs to be uniform across
+    // meshblocks that meshblock packs will work
+    fsize += v->NumComponents();
   }
 
   // make the outer view
   ViewOfParArrays<T> cv("MakeFluxPack::cv", vsize);
   ViewOfParArrays<T> f1("MakeFluxPack::f1", fsize);
   ViewOfParArrays<T> f2("MakeFluxPack::f2", fsize);
   ViewOfParArrays<T> f3("MakeFluxPack::f3", fsize);
-  ParArray1D<int> sparse_assoc("MakeFluxPack::sparse_assoc", vsize);
+  ParArray1D<int> sparse_id("MakeFluxPack::sparse_id", vsize);
   ParArray1D<int> vector_component("MakeFluxPack::vector_component", vsize);
```

Comment:
I'm pretty sure `sparse_id` and `sparse_assoc` are *not* the same. The former is the id of the variable. The latter is the variable associated with it for the purposes of, e.g., fluxes. @jdolence can say more.

Reply:
Hmm, so previously, `FillVarView` did this:
```c++
// ...
auto host_sp = Kokkos::create_mirror_view(Kokkos::HostSpace(), sparse_assoc);
// ...
for (const auto v : vars) {
  int sparse_id = v->metadata().GetSparseId();
  // ...
  host_sp(vindex) = sparse_id;
  // ...
}
// ...
Kokkos::deep_copy(sparse_assoc, host_sp);
```
So it seems that `sparse_assoc` was just filled with `sparse_id = v->metadata().GetSparseId()`, so it is the pure sparse ID. Unless this was a mistake...

Reply:
Huh. This might be related to the other discussion about location in the pack vs global id. Let's leave your changes in place for now and address in the next PR.

Unless you object @jdolence ?